File: flink-jobs/helper/src/main/java/ch/rwicht/rbd/helper/JSONKeyValueSerializationSchema.java
package ch.rwicht.rbd.helper;

import org.apache.flink.annotation.PublicEvolving;
import org.apache.flink.shaded.jackson2.com.fasterxml.jackson.core.JsonProcessingException;
import org.apache.flink.shaded.jackson2.com.fasterxml.jackson.databind.ObjectMapper;
import org.apache.flink.shaded.jackson2.com.fasterxml.jackson.databind.ObjectWriter;
import org.apache.flink.shaded.jackson2.com.fasterxml.jackson.databind.node.ObjectNode;
import org.apache.flink.streaming.connectors.kafka.KafkaSerializationSchema;
import org.apache.kafka.clients.producer.ProducerRecord;

import javax.annotation.Nullable;

@PublicEvolving
public class JSONKeyValueSerializationSchema implements KafkaSerializationSchema<ObjectNode> {
    private ObjectWriter writer;
    private String topic;

    public JSONKeyValueSerializationSchema(String topic) {
        this.topic = topic;
    }

    @Override
    public ProducerRecord<byte[], byte[]> serialize(ObjectNode jsonNodes, @Nullable Long aLong) {
        if (this.writer == null) {
            this.writer = new ObjectMapper().writer();
        }

        byte[] bytes = new byte[0];
        try {
            bytes = this.writer.writeValueAsBytes(jsonNodes);
        } catch (JsonProcessingException e) {
            e.printStackTrace();
        }
        return new ProducerRecord<>(topic, bytes);
    }
}

File: flink-jobs/access-parser/src/main/scala/ch/rwicht/rbd/schemas/JSONKeyValueSerializationSchema.scala
package ch.rwicht.rbd.schemas

import java.lang

import org.apache.flink.shaded.jackson2.com.fasterxml.jackson.databind.ObjectMapper
import org.apache.flink.streaming.connectors.kafka.KafkaSerializationSchema
import org.apache.kafka.clients.producer.ProducerRecord

class JSONKeyValueSerializationSchema[ObjectNode](topic: String) extends KafkaSerializationSchema[ObjectNode] {
  override def serialize(t: ObjectNode, aLong: lang.Long): ProducerRecord[Array[Byte], Array[Byte]] = {
    val writer = new ObjectMapper().writer()
    val bytes = writer.writeValueAsBytes(t)
    new ProducerRecord[Array[Byte], Array[Byte]](topic, bytes)
  }
}
